app/services/organization_update_service.rb

```diff
@@ -1,27 +1,52 @@
 class OrganizationUpdateService
-  # @param organization [Organization]
-  # @param params [ActionDispatch::Http::Parameters]
-  # @return [Boolean]
-  def self.update(organization, params)
-    result = organization.update(params)
-    return false unless result
-
-    update_partner_flags(organization)
-    true
-  end
-
-  # @param organization [Organization]
-  def self.update_partner_flags(organization)
-    fields = %i[
+  class << self
+    FIELDS = %i[
       enable_child_based_requests
       enable_individual_requests
       enable_quantity_based_requests
     ]
-    fields.each do |field|
-      next unless organization.saved_change_to_attribute?(field)
-      organization.partners.map(&:profile).each do |profile|
-        profile.update!(field => organization.send(field))
+
+    # @param organization [Organization]
+    # @param params [ActionDispatch::Http::Parameters]
+    # @return [Boolean]
+    def update(organization, params)
+      return false unless valid?(organization, params)
+
+      result = organization.update(params)
+      return false unless result
+
+      update_partner_flags(organization)
+      true
+    end
+
+    # @param organization [Organization]
+    def update_partner_flags(organization)
+      FIELDS.each do |field|
+        next if organization.send(field)
+        organization.partners.map(&:profile).each do |profile|
+          profile.update!(field => organization.send(field))
+        end
       end
     end
+
+    private
+
+    def valid?(organization, params)
+      return true unless organization.partners.any?
+
+      disable_params = FIELDS.select { |field| params[field] == false }
+
+      organization.partners.each do |partner|
+        return false if disables_all_partner_fields?(partner, disable_params)
+      end
+
+      true
+    end
+
+    def disables_all_partner_fields?(partner, disable_params)
+      enabled_fields = FIELDS.select { |field| partner.profile.send(field) == true }
+
+      enabled_fields.all? { |field| disable_params.include?(field) }
```

Comment:
There are a lot of `fields` and `params` flying around here and in general I find this code hard to follow. I think we need to have a more hardcoded workflow for the given AC's to make it clearer what's supposed to happen when. I *think* this is saying that we shouldn't allow the update to proceed if by doing so we would end up disabling all partner fields at once, but can't we do this a lot simpler just by checking to see if all request types end up being false on the organization, and not having to check the partners at all?

Reply:
Agreed, I wasn't super happy with how readable this is. I'd love to make it better 💪

> but can't we do this a lot simpler just by checking to see if all request types end up being false on the organization, and not having to check the partners at all?

As I understand it: we need to check all partners for a given organisation because of AC2 in the PR description - essentially, we don't want to disable a request type in a partner if it is the only one enabled. Therefore, validation has to happen at the partner level.

I'll try moving some logic to the Organisation and Partner classes - that wouldn't solve the issue of the logic itself per se, but it would make this service class more readable (hopefully! 🤞 )

Reply:
Just realised: moving logic to other classes isn't the best option because we'd also then have to move the constant `FIELDS` across as well. Seeing as this service class is the only class to be concerned with `FIELDS`, I'll retain the logic inside this class